### PR TITLE
CurlHttpClient: Add seek support.

### PR DESCRIPTION
**Pull request by whoenig**
This fixes issue #794.

### CODE REVIEW

**Comment by marcomagdy** [comment]
Will run this through our CI first and merge if it passes.
Thank you.

File: aws-cpp-sdk-core/source/http/curl/CurlHttpClient.cpp [hunk 1]
```diff
@@ -590,3 +592,41 @@ size_t CurlHttpClient::ReadBody(char* ptr, size_t size, size_t nmemb, void* user
...
+            break;
+        case SEEK_END:
+            dir = std::ios_base::end;
+            break;
+        default:
+            return CURL_SEEKFUNC_FAIL;
+    }
+
+    ioStream->clear();
+    ioStream->seekg(offset, dir);
```
Comment (marcomagdy) [inline aws-cpp-sdk-core/source/http/curl/CurlHttpClient.cpp]: You need to check if `seekg` failed and return `CURL_SEEKFUNC_CANTSEEK` Some streams are not seek-able, for example crypto streams using GCM and GMAC

Comment (whoenig) [reply aws-cpp-sdk-core/source/http/curl/CurlHttpClient.cpp]: Good point - thanks! I updated the PR accordingly.

File: aws-cpp-sdk-core/source/http/curl/CurlHttpClient.cpp [hunk 2]
```diff
@@ -590,3 +592,44 @@ size_t CurlHttpClient::ReadBody(char* ptr, size_t size, size_t nmemb, void* user
     return 0;
 }
 
+size_t CurlHttpClient::SeekBody(void* userdata, curl_off_t offset, int origin)
+{
+    CurlReadCallbackContext* context = reinterpret_cast<CurlReadCallbackContext*>(userdata);
```
Comment (singku) [inline aws-cpp-sdk-core/source/http/curl/CurlHttpClient.cpp]: How about 
```
if (!userdata)
{
    return CURL_SEEKFUNC_FAIL;
}
```
then
```
CurlReadCallbackContext* context = reinterpret_cast<CurlReadCallbackContext*>(userdata);

```

**Comment by marcomagdy** [comment]
I merged the PR to our internal repo.
It will make it to the public repo with the next batch of service patches.

Thanks for the contribution

**Comment by marcomagdy** [comment]
This has been merged in merged in 3fb3bcd912570f6c257c4601423dd8eb438f110c

**Pull request closed without merging**